src/abstractdataframe/abstractdataframe.jl

```diff
@@ -983,35 +983,88 @@ julia> vcat(df1, df2)
-    intersectunique = intersect(uniqueheaders...)
-    coldiff = setdiff(unionunique, intersectunique)
-
-    if !isempty(coldiff)
+    # array of all headers
+    @show allheaders = map(names, dfs)
+    # unique arrays of all headers
+    @show uniqueheaders = unique(allheaders)
+    # Array of all the unique headers
+    @show unionunique = union(uniqueheaders...)
+    # Intersection of all unique headers 
+    @show intersectunique = intersect(uniqueheaders...)
+    # get the elements that are not present in everything
+    @show coldiff = setdiff(unionunique, intersectunique)
+
+    @show keep
+    if (widen == false) && !isempty(coldiff) && keep == nothing
         # if any DataFrames are a full superset of names, skip them
         filter!(u -> Set(u) != Set(unionunique), uniqueheaders)
         estrings = Vector{String}(undef, length(uniqueheaders))
         for (i, u) in enumerate(uniqueheaders)
-            matching = findall(h -> u == h, allheaders)
-            headerdiff = setdiff(coldiff, u)
-            cols = join(headerdiff, ", ", " and ")
-            args = join(matching, ", ", " and ")
+            @show matching = findall(h -> u == h, allheaders)
+            @show headerdiff = setdiff(coldiff, u)
+            @show cols = join(headerdiff, ", ", " and ")
+            @show args = join(matching, ", ", " and ")
             estrings[i] = "column(s) $cols are missing from argument(s) $args"
         end
         throw(ArgumentError(join(estrings, ", ", ", and ")))
     end
 
-    header = allheaders[1]
+    # TODO: make sure that `keep` throws a good error if 
+    # it a) isn't in `allheaders` or b) isn't a subset of `unionunique`
+    header = (keep == nothing) ? unionunique : keep 
+
+    if keep == nothing 
+        # Make the order of the names match the order of the dataframes inputted
+        header = let unionunique = unionunique, allheaders = allheaders
+            t = [filter(h -> h in unionunique, head) for head in allheaders]
+            reduce((a, b) -> [a; setdiff(b, a)], t)
+        end
+    else 
+        header = keep 
+    end
+
     length(header) == 0 && return DataFrame()
     cols = Vector{AbstractVector}(undef, length(header))
     for (i, name) in enumerate(header)
-        data = [df[name] for df in dfs]
+        # TODO: replace with commented out code after getindex deprecation
+        # data = [df[name] for df in dfs]
+        # the code below assumes that only DataFrame and SubDataFrame
+        # are subtypes of AbstractDataFrame
+        # it should be removed ASAP after deprecation
+        data = map(dfs) do df
+            if df isa DataFrame 
+                if haskey(df, name)
+                    return df[name] 
+                else
+                    # TODO: make this more efficient by not creating a 
+                    # full array of missing values. Instead, implement
+                    # this in the copyto! stage
+                    return fill(missing, nrow(df))
```

Comment:
Sorry for the delay, I had missed the notification. `RepeatedVector` can probably help here (if using a generator isn't enough).

Reply:
ugh why did I not think of a generator. It's fixed now.

Reply:
It turns out that a generator doesn't infer eltype correclty. Columns are of type `Any` rather than `Union{T, Missing}`. 

`RepeatedVector` doesn't work because it's not defined at this point in the code, so we can't use it here. (Should we have a place for all struct defintions?)

Let me know what to do next. I think there is an issue for this generator issue but I can't find it.

Reply:
Ah. Maybe with `Iterators.repeated`? (`eltype` won't be defined for generators since it needs inference.)